### PR TITLE
Plugins: move "Upload Plugin" button to the navigation bar

### PR DESCRIPTION
**Pull request by jsnajdr**
Use the `HeaderButton` component for the "Upload Plugin" button and move the button from the list header to the navigation bar.

The "Manage", "Add Plugin" and "Upload Plugin" now share the same location and style.

I also cleaned up the `HeaderButton` styling in the second commit. See the commit message for details. Cc @iamtakashi 

There's one more improvement I'd like to do in this PR before it's ready to merge: fine-tune the horizontal layout of the navigation bar. Using a `@media` breakpoint doesn't look like a good solution any more:

At width of 680px, the `SectionNav` is far too narrow and the bar should break into two lines:
section-nav-at-680-px

But at a **smaller** width of 650px, the `SectionNav` fits into available space just fine:
section-nav-at-650-px

I'd like to try out another approach: use `flex-wrap: wrap` to optionally wrap the bar into two lines, and specify the point where this happens by setting `min-width` on the `SectionNav`.

This PR continues the Plugins improvements work that was started in #17537.

### CODE REVIEW

**Comment by iamtakashi** [comment]
@jsnajdr What I did for the narrower viewport was to drop the buttons to next row. See the production. And the empty space is for the upload button.

screen shot 2017-09-05 at 15 27 09

**Comment by jsnajdr** [comment]
> What I did for the narrower viewport was to drop the buttons to next row. See the production. And the empty space is for the upload button.

@iamtakashi Yes, and that code is still in place. But the next row is displayed only when the viewport width is smaller than 480px. In some cases, even 650px is not big enough.

**Comment by iamtakashi** [comment]
OK, then let's do the following.

column -> row -> column -> row

**Comment by iamtakashi** [comment]
Also the cloud upload needs a custom positioning there. Adding the following should make it nicer.

```
.plugins__header-buttons .button .gridicon.gridicons-cloud-upload {
	top: 6px;
}
```

**Comment by iamtakashi** [comment]
Another thing I spotted: It looks like the hover style for the buttons seems disappeared at some point. Both buttons should have `background: $gray-light` for `:hover`.

**Comment by tyxla** [comment]
> Also the cloud upload needs a custom positioning there. Adding the following should make it nicer.
>
> ```
> .plugins__header-buttons .button .gridicon.gridicons-cloud-upload {
>	top: 6px;
>}
>```

I think it should be good as-is, see this PR and the explanation on why: https://github.com/Automattic/wp-calypso/pull/17630#issuecomment-326256209.

**Comment by iamtakashi** [comment]
> I think it should be good as-is, see this PR and the explanation on why: #17630 (comment).

@tyxla The change you're referring to seems to change the positioning in compact buttons only. Therefore, we still need the custom positioning here.

**Comment by jsnajdr** [comment]
I pushed a few commits that do the following:
- tweak the position of the cloud icon in `HeaderButton`
- introduce the `:hover` style for `HeaderButton` -- apparently one of the patches that got lost during one of @sirbrillig's rebases ;)
- add "Upload Plugin" button also to the Plugin Browser page

I think we are ready for testing and merge -- Cc @iamtakashi @seear @tyxla 

The only thing that remains is to solve the `SectionNav` overlap when the viewport is too narrow. I'd like to avoid introducing mutliple breakpoints that would switch from row to column view and back again -- I'm working on a more elegant solution based on `flex-wrap`. But I need more time to finish it and don't want to delay merging of all the useful improvements that are already ready here.

File: client/my-sites/plugins/plugins-browser/index.jsx [hunk 1]
```diff
@@ -311,14 +339,18 @@ const PluginsBrowser = React.createClass( {
 		}
 
 		const navigation = this.props.category ? this.getNavigationBar() : this.getSearchBar();
-		const manageButton = this.shouldShowManageButton() && this.getManageButton();
 
+		/* eslint-disable wpcalypso/jsx-classname-namespace */
```
Comment (seear) [inline client/my-sites/plugins/plugins-browser/index.jsx]: There might be a way to organise things so that the header buttons become a separate component, with the actual buttons passed in as children, or maybe props to control which buttons to show. That would take care of the lint warning and a little bit of duplication. Would be for a different PR, though.

File: client/my-sites/plugins/plugins-browser/index.jsx [hunk 2]
```diff
@@ -287,7 +288,18 @@ const PluginsBrowser = React.createClass( {
 		);
 	},
 
-	getManageButton() {
+	shouldShowManageButton() {
+		if ( this.props.isJetpackSite ) {
+			return true;
+		}
+		return ( ! this.props.selectedSiteId && this.props.hasJetpackSites );
```
Comment (seear) [inline client/my-sites/plugins/plugins-browser/index.jsx]: Loading with empty redux state on a simple site, I get a flash of the manage button. Possible that the logic could be tweaked here to avoid that.

Comment (jsnajdr) [reply client/my-sites/plugins/plugins-browser/index.jsx]: I have trouble reproducing this and I don't immediately see any flaw in the logic. Could you record a video? Maybe that will give me a clue.

I'll merge this PR now and will solve any remaining problems separately.

**Comment by tyxla** [comment]
Looks and works as a charm!

File: client/my-sites/plugins/plugins-browser/index.jsx (hunk 2 above)
Comment (seear) [reply client/my-sites/plugins/plugins-browser/index.jsx]: Here it is:

manage_flash

`hasJetpackSites` gets populated before `selectedSiteId`. I'm not too sure how this would be fixed. Possible that  `getSelectedSiteId()` should distinguish between _no site_ selected and _not yet known_. Currently it returns `null` for both.

Comment (seear) [reply client/my-sites/plugins/plugins-browser/index.jsx]: Seems way beyond the scope of current work, anyway.
